src/lib.rs

```diff
@@ -64,13 +64,40 @@ pub fn compile(
     debug_ast: bool,
     debug_ir: bool,
 ) -> Result<Product, Error> {
-    let lexer = Lexer::new(filename, buf.chars(), debug_lex);
-    let parser = Parser::new(lexer, debug_ast)?;
+    use data::error::Error as SourceError;
+    let mut lexer = Lexer::new(filename, buf.chars(), debug_lex);
+
+    let mut lex_errs = VecDeque::new();
+    let first = loop {
+        match lexer.next() {
+            Some(Ok(token)) => break token,
+            Some(Err(err)) => lex_errs.push_back(err),
+            None => {
+                if lex_errs.is_empty() {
+                    lex_errs.push_back(CompileError::new(
+                        SourceError::EmptyProgram,
+                        lexer.location(),
+                    ));
+                }
+                return Err(Error::Source(lex_errs));
+            }
+        }
+    };
```

Comment:
This situation seems pretty complicated for a public API. Maybe there should be a helper function?

Reply:
Sure, I can add another `.collect_results()`.